composer/trainer/trainer.py

```diff
@@ -680,25 +685,44 @@ def _train_loop(self) -> None:
                         "trainer/global_step": self.state.step,
                         "trainer/batch_idx": self.state.timer.batch_in_epoch.value,
                     })
+                    # Rerun train_batch with smaller mini-batches if we hit CUDA out of memory
+                    rerun_train_batch = True
                     total_loss = None
-                    microbatches = self._train_data_spec.split_batch(state.batch, state.grad_accum)
-                    if self.deepspeed_enabled:
-                        total_loss = self._train_batch(microbatches)
-                    elif self._use_closures():
-                        for optimizer in state.optimizers:
-                            if use_grad_scaling:
-                                total_loss = state.scaler.step(
-                                    optimizer, closure=lambda **kwargs: self._train_batch(microbatches, **kwargs))
+                    while rerun_train_batch:
```

Comment:
This should be tested on the existing composer benchmarks (see the yaml files in composer/yamls for models to run).

In general this will work in the majority of cases, however there are a few problematic edge cases we want to verify: :
- Sometimes memory usage differs between ranks - will this result in different ranks having different values for  grad_accum? The majority of the time it will not but we want to make sure it doesn't cause problems in cases where we are very close to max memory usage (for example, different processes having different values for grad accum will result in our profiler results being different per process).
- If one process crashes but others do not, will it result in the `loss.backwards()` hanging?


Just as a point of consideration - one other way to implement this is to only have rank0 do the grad_accum search and then broadcast the result to other ranks